### PR TITLE
http: sending the original url scheme (or best guess) upstream.

### PR DESCRIPTION
**Pull request by alyssawilk**
This replaces prior logic where the :scheme header was consistently overwritten based on the encryption level of the upstream connection.

Risk Level: High (l7 change)
Testing: new integration tests, unit tests
Docs Changes: api docs updated
Release Notes: inline
Runtime guard: envoy.reloadable_features.preserve_downstream_scheme
Part of #14587

### CODE REVIEW

File: source/common/router/router.cc [hunk 1]
```diff
@@ -80,8 +80,23 @@ uint64_t FilterUtility::percentageOfTimeout(const std::chrono::milliseconds resp
...
+  if (Runtime::runtimeFeatureEnabled("envoy.reloadable_features.preserve_downstream_scheme")) {
+    if (Http::HeaderUtility::schemeIsValid(headers.getSchemeValue())) {
+      return;
+    }
+    if (Http::HeaderUtility::schemeIsValid(headers.getForwardedProtoValue())) {
+      headers.setScheme(headers.getForwardedProtoValue());
+      return;
+    }
+  }
+  bool transport_secure =
```
Comment (mattklein123) [inline source/common/router/router.cc:94]: nit: const downstream_transport_secure

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Thanks LGTM modulo nit. I think it will be a miracle if this doesn't break someone but here's hoping. :crossed_fingers: 

/wait

**Comment by alyssawilk** [comment]
oh yeah, I meant to move the warning to the top because of risk.  will dot that now.
also making transport_secure const, but not renaming since the direction still depends on the runtime value.

**Comment by alyssawilk** [comment]
sorry, lost my LGTM with tidy and accidentally force pushed a delta but I think your review state should all be there

**Review by mattklein123 — APPROVED** [review]
Thanks!

**Pull request merged**